portal-bridge/src/gossip.rs

```diff
@@ -0,0 +1,146 @@
+use crate::stats::{BeaconSlotStats, HistoryBlockStats, StatsReporter};
+use ethportal_api::jsonrpsee::core::Error;
+use ethportal_api::types::portal::{ContentInfo, TraceGossipInfo};
+use ethportal_api::{
+    BeaconContentKey, BeaconContentValue, BeaconNetworkApiClient, HistoryContentKey,
+    HistoryContentValue, HistoryNetworkApiClient, OverlayContentKey, PossibleBeaconContentValue,
+    PossibleHistoryContentValue,
+};
+use jsonrpsee::http_client::HttpClient;
+use std::sync::{Arc, Mutex};
+use tokio::time::{sleep, Duration};
+use tracing::{debug, warn};
+
+const GOSSIP_RETRY_COUNT: u64 = 3;
+const RETRY_AFTER: Duration = Duration::from_secs(15);
+
+/// Gossip any given content key / value to the history network.
+pub async fn gossip_beacon_content(
+    portal_clients: Arc<Vec<HttpClient>>,
+    content_key: BeaconContentKey,
+    content_value: BeaconContentValue,
+    slot_stats: Arc<Mutex<BeaconSlotStats>>,
+) -> anyhow::Result<()> {
+    let mut results: Vec<Result<(Vec<TraceGossipInfo>, u64), Error>> = vec![];
+    for client in portal_clients.as_ref() {
+        let client = client.clone();
+        let content_key = content_key.clone();
+        let content_value = content_value.clone();
+        let result = tokio::spawn(beacon_trace_gossip(client, content_key, content_value)).await?;
+        results.push(result);
+    }
+    if let Ok(mut data) = slot_stats.lock() {
+        data.update(content_key, results.into());
+    } else {
+        warn!("Error updating beacon gossip stats. Unable to acquire lock.");
+    }
+    Ok(())
+}
+
+async fn beacon_trace_gossip(
+    client: HttpClient,
+    content_key: BeaconContentKey,
+    content_value: BeaconContentValue,
+) -> Result<(Vec<TraceGossipInfo>, u64), Error> {
+    let mut retry_count = 0;
+    let mut traces = vec![];
+    while retry_count < GOSSIP_RETRY_COUNT {
+        let result = BeaconNetworkApiClient::trace_gossip(
+            &client,
+            content_key.clone(),
+            content_value.clone(),
+        )
+        .await;
+        // check if content was successfully transferred to at least one peer on network
+        if let Ok(trace) = result {
+            traces.push(trace.clone());
+            if !trace.transferred.is_empty() {
+                return Ok((traces, retry_count));
+            }
+        }
+        // if not, make rfc request to see if data is available on network
+        let result =
+            BeaconNetworkApiClient::recursive_find_content(&client, content_key.clone()).await;
+        if let Ok(PossibleBeaconContentValue::ContentPresent(_)) = result {
```

Comment:
Beacon should return `ContentInfo::Content` (same as the history below). They have identical jsonr-rpc implementations:  [beacon](https://github.com/ethereum/trin/blob/092427a4f0babf6a8bc3423de84a89b4f244d0cd/trin-beacon/src/jsonrpc.rs#L125) | [history](https://github.com/ethereum/trin/blob/092427a4f0babf6a8bc3423de84a89b4f244d0cd/trin-history/src/jsonrpc.rs#L126).

Reply:
Agreed that the subnetwork returns the same type, but the rpc deserialization of the response are different for the two endpoints: [beacon](https://github.com/ethereum/trin/blob/092427a4f0babf6a8bc3423de84a89b4f244d0cd/rpc/src/beacon_rpc.rs#L146C29-L146C29) | [history](https://github.com/ethereum/trin/blob/092427a4f0babf6a8bc3423de84a89b4f244d0cd/rpc/src/history_rpc.rs#L122C20-L122C31)